File: leetcode/0031_next_permutation.cpp
// https://leetcode.com/problems/permutations

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> r;
        permute(nums, 0, r);
        return r;
    }
    
    void permute(vector<int>& nums, int iStart, vector<vector<int>>& r)
    {
        if (iStart >= nums.size())
        {
            r.push_back(nums);
            return;
        }
        
        for (int i = iStart; i < nums.size(); ++i)
        {
            swap(nums[iStart], nums[i]);
            permute(nums, iStart + 1, r);
            if (i < nums.size() - 1)
                reverse(nums.begin() + iStart + 1, nums.end());
        }
    }
};

File: leetcode/0031_next_permutation.cs
// https://leetcode.com/problems/next-permutation/description/

public class Solution {
    public void NextPermutation(int[] nums) {
        if (nums.Length < 2) return;

        int i = nums.Length - 2;

        for (; i >= 0 && nums[i] >= nums[i+1]; --i)
            ;

        int j = i + 1;
        for (int k = nums.Length - 1; j < k; ++j, --k)
        {
            var e = nums[j];
            nums[j] = nums[k];
            nums[k] = e;
        }
        
        if (i < 0) return;

        j = i + 1;
        for (; j < nums.Length && nums[j] <= nums[i]; ++j)
            ;
        
        if (j >= nums.Length) return;
        
        var t = nums[i];
        nums[i] = nums[j];
        nums[j] = t;
    }
}
